### PR TITLE
feat: allow exiting offers

### PR DESCRIPTION
**Pull request by samsiegart**
fixes: https://github.com/Agoric/wallet-app/issues/61

- Adds exit button to currently-seated offers
- Adds refunded status for offers with 0 wants satisfied

Tested making liquidation bids from CLI and exiting them from UI

image

image

### CODE REVIEW

File: wallet/src/service/Offers.ts [hunk 1]
```diff
@@ -206,45 +207,95 @@ export const getOfferService = (
     return signSpendAction(offer.spendAction);
   };
 
-  const cancelOffer = _id => {
-    console.log('TODO: cancel offer');
+  const cancelOffer = async (id: number) => {
```
Comment (turadg) [inline wallet/src/service/Offers.ts:210]: let's be consistent in terminology:
```suggestion
  const tryExitOffer = async (id: number) => {
```
"cancel" suggests that it'll always work

**Review by turadg — COMMENTED** [review]
Will review again after the contract lands

File: wallet/src/components/Offer.tsx [hunk 2]
```diff
@@ -141,11 +134,14 @@ const OfferWithoutContext = ({
         color={statusColors[status]}
         label={statusText[status]}
       />
-      <span className="Date text-gray">{formatDateNow(date)}</span>
       <div className="OfferOrigin" style={{ wordBreak: 'break-word' }}>
         <PetnameSpan name={sourceDescription} />
-        <i> via </i>
-        <span className="Blue">{dappOrigin || origin}</span>
+        {(dappOrigin || origin) && (
```
Comment (turadg) [inline wallet/src/components/Offer.tsx:139]: consider `const actualOrigin = dappOrigin || origin` so the condition doesn't need to be repeated on line 142

File: wallet/src/service/Offers.ts [hunk 3]
```diff
@@ -207,45 +208,95 @@ export const getOfferService = (
...
+          const value = String(entry.value);
+          return [kw, { pursePetname, value }];
+        }),
+      );
+
+    for await (const pendingOffers of makeAsyncIterableFromNotifier(
+      pendingOffersNotifier,
+    )) {
+      console.log('pending offers', pendingOffers);
+      pendingOffers?.forEach(([_, o]) => {
```
Comment (turadg) [inline wallet/src/service/Offers.ts:276]: prefer control structures when available. `for` loop allows `continue` e.g.,
```suggestion
      if (!pendingOffers) continue;
      for (const [_, o] of pendingOffers) {
```

File: wallet/src/components/tests/Offer.test.jsx [hunk 4]
```diff
@@ -201,8 +190,8 @@ test('renders the controls', () => {
   expect(controls.find(Chip).at(1).text()).toContain('Decline');
 });
 
-test.skip('renders the exit button while pending', () => {
-  const pendingOffer = { ...offer };
+test('renders the exit button while seated', () => {
```
Comment (turadg) [inline wallet/src/components/tests/Offer.test.jsx:193]: glad to see these tests

**Review by turadg — APPROVED** [review]

File: wallet/src/service/Offers.ts (hunk 3 above)
Comment (samsiegart) [reply wallet/src/service/Offers.ts]: Done

File: wallet/src/components/Offer.tsx (hunk 2 above)
Comment (samsiegart) [reply wallet/src/components/Offer.tsx]: Nice done.

**Pull request merged**